src/Model/TransportModel/tsp-adv.f90

```diff
@@ -127,6 +130,71 @@ subroutine adv_ar(this, dis, ibound)
     return
   end subroutine adv_ar
 
+  !> @brief  Calculate maximum time step length
+  !!
+  !!  Return the largest time step that meets stability constraints
+  !<
+  subroutine adv_dt(this, dtmax, msg, thetam)
+    ! dummy
+    class(TspAdvType) :: this !< this instance
+    real(DP), intent(out) :: dtmax !< maximum allowable dt subject to stability constraint
+    character(len=*), intent(inout) :: msg !< package/cell dt constraint message
+    real(DP), dimension(:), intent(in) :: thetam !< porosity
+    ! local
+    integer(I4B) :: n
+    integer(I4B) :: m
+    integer(I4B) :: ipos
+    integer(I4B) :: nrmax
+    character(len=LINELENGTH) :: cellstr
+    real(DP) :: dt
+    real(DP) :: qmax
+    real(DP) :: qnmsumpos
+    real(DP) :: qnmsumneg
+    real(DP) :: qnm
+    real(DP) :: cell_volume
+    dtmax = DNODATA
+    nrmax = 0
+    msg = ''
+
+    ! If ats_percel not specified by user, then return without making
+    ! the courant time step calculation
+    if (this%ats_percel == DNODATA) then
+      return
+    end if
+
+    ! Calculate time step lengths based on stability constraint for each cell
+    ! and store the smallest one
+    do n = 1, this%dis%nodes
+      if (this%ibound(n) == 0) cycle
+      qnmsumneg = DZERO
+      qnmsumpos = DZERO
+      do ipos = this%dis%con%ia(n) + 1, this%dis%con%ia(n + 1) - 1
+        if (this%dis%con%mask(ipos) == 0) cycle
+        m = this%dis%con%ja(ipos)
+        if (this%ibound(m) == 0) cycle
+        qnm = this%fmi%gwfflowja(ipos)
+        if (qnm < DZERO) then
+          qnmsumneg = qnmsumneg - qnm
+        else
+          qnmsumpos = qnmsumpos + qnm
+        end if
+      end do
+      qmax = max(qnmsumneg, qnmsumpos)
```

Comment:
Isn't this condition to restrictive? Wouldn't it sufficive to take the biggest abs(qnm) as that is the velocity that dominates the propagation of information?

Reply:
I don't think that's generally true for unstructured grids.  I think the best we can do here is look at total inflows and total outflows and use the largest of the two for the constraint calculation.  Let me know if you have ideas for a better way to handle this for unstructured grids.  I suspect it's possible, but I think it would take some geometric processing to get the proper cell length in three dimensions.